Userland/Libraries/LibWeb/WebAudio/AudioBuffer.cpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024, Shannon Booth <shannon@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibJS/Runtime/Completion.h>
+#include <LibJS/Runtime/Realm.h>
+#include <LibJS/Runtime/TypedArray.h>
+#include <LibJS/Runtime/VM.h>
+#include <LibWeb/Bindings/Intrinsics.h>
+#include <LibWeb/WebAudio/AudioBuffer.h>
+#include <LibWeb/WebAudio/BaseAudioContext.h>
+#include <LibWeb/WebIDL/DOMException.h>
+
+namespace Web::WebAudio {
+
+JS_DEFINE_ALLOCATOR(AudioBuffer);
+
+WebIDL::ExceptionOr<JS::NonnullGCPtr<AudioBuffer>> AudioBuffer::construct_impl(JS::Realm& realm, AudioBufferOptions const& options)
+{
+    auto& vm = realm.vm();
+
+    // 1. If any of the values in options lie outside its nominal range, throw a NotSupportedError exception and abort the following steps.
+    TRY(BaseAudioContext::verify_audio_options_inside_nominal_range(realm, options.number_of_channels, options.length, options.sample_rate));
+
+    // 2. Let b be a new AudioBuffer object.
+    // 3. Respectively assign the values of the attributes numberOfChannels, length, sampleRate of the AudioBufferOptions passed in the
+    //    constructor to the internal slots [[number of channels]], [[length]], [[sample rate]].
+    auto buffer = vm.heap().allocate<AudioBuffer>(realm, realm, options);
+
+    // 4. Set the internal slot [[internal data]] of this AudioBuffer to the result of calling CreateByteDataBlock([[length]] * [[number of channels]]).
+    Vector<JS::NonnullGCPtr<JS::Float32Array>> channels;
+    channels.ensure_capacity(options.number_of_channels);
+
+    for (WebIDL::UnsignedLong i = 0; i < options.number_of_channels; ++i)
+        channels.append(TRY(JS::Float32Array::create(realm, options.length)));
```

Comment:
Is this safe? Is it possible for the `JS::Float32Array` allocation at, say, index 2 to cause a GC and free the allocations from indices 0 and 1? Maybe we should just work directly on `m_channels`, which will be visited:

```c++
    buffer->m_channels.ensure_capacity(options.number_of_channels);

    for (WebIDL::UnsignedLong i = 0; i < options.number_of_channels; ++i)
        buffer->m_channels.append(TRY(JS::Float32Array::create(realm, options.length)));
```